File: 238. Product of Array Except Self/productExceptSelf.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        length = len(nums)
        ans = [0] * length
        
        for i in range(length):
            if i == 0:
                ans[i] = 1
            else:
                ans[i] = ans[i - 1] * nums[i - 1]
        
        rightProduct = 1
        for i in range(length - 2, -1, -1):
            rightProduct *= nums[i + 1]
            ans[i] *= rightProduct
        
        return ans

File: 238. Product of Array Except Self/productExceptSelf.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int length = nums.length;
        int[] left = new int[length];
        int[] right = new int[length];
        int leftProduct = 1, rightProduct = 1;
        
        for (int i = 0; i < length; i++) {
            left[i] = leftProduct;
            leftProduct *= nums[i];
        }
        for (int i = length - 1; i >= 0; i--) {
            right[i] = rightProduct;
            rightProduct *= nums[i];
        }

        for (int i = 0; i < length; i++) {
            nums[i] = left[i] * right[i];
        }
        return nums;

    }
}